platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/migratable/ProfileMigratable.java

```diff
@@ -0,0 +1,251 @@
+  public String getTitle() {
+    return "Profile Migratable";
+  }
+
+  @Override
+  public String getDescription() {
+    return "Exports the state of applications, features, and bundles";
+  }
+
+  @Override
+  public String getOrganization() {
+    return "Codice";
+  }
+
+  @Override
+  public void doExport(ExportMigrationContext context) {
+    context
+        .getEntry(ProfileMigratable.PROFILE_PATH)
+        .store(
+            (r, os) ->
+                ProfileMigratable.MAPPER.writeValue(
+                    os,
+                    ImmutableMap.of( //
+                        ProfileMigratable.PROFILE_APPS,
+                        appMgr.exportApplications(),
+                        ProfileMigratable.PROFILE_FEATURES,
+                        featureMgr.exportFeatures(),
+                        ProfileMigratable.PROFILE_BUNDLES,
+                        bundleMgr.exportBundles())));
+  }
+
+  @Override
+  public void doImport(ImportMigrationContext context) {
+    try {
+      context.getEntry(ProfileMigratable.PROFILE_PATH).restore(this::restore);
+    } catch (MigrationException e) { // don't want to abort the import; just record the error
+      context.getReport().record(e);
+    }
+  }
+
+  @SuppressWarnings(
+      "squid:S2093" /* try-with-resource will throw IOException with InputStream and we do not care to get that exception */)
+  private boolean restore(MigrationReport report, Optional<InputStream> ois) throws IOException {
+    final InputStream is = ois.orElse(null);
+
+    if (is == null) {
+      throw new MigrationException("Import error: missing exported profile information.");
+    }
+    final Map<String, Object> pinfo;
+
+    try {
+      pinfo =
+          ProfileMigratable.MAPPER
+              .parser()
+              .parseMap(IOUtils.toString(is, ProfileMigratable.DEFAULT_CHARSET));
+    } finally {
+      IOUtils.closeQuietly(is);
+    }
+    final List<JsonApplication> japps =
+        JsonUtils.getListFrom(pinfo, ProfileMigratable.PROFILE_APPS, true)
+            .stream()
+            .map(JsonUtils::convertToMap)
+            .map(JsonApplication::new)
+            .collect(Collectors.toList());
+    final List<JsonFeature> jfeatures =
+        JsonUtils.getListFrom(pinfo, ProfileMigratable.PROFILE_FEATURES, true)
+            .stream()
+            .map(JsonUtils::convertToMap)
+            .map(JsonFeature::new)
+            .collect(Collectors.toList());
+    final List<JsonBundle> jbundles =
+        JsonUtils.getListFrom(pinfo, ProfileMigratable.PROFILE_BUNDLES, true)
+            .stream()
+            .map(JsonUtils::convertToMap)
+            .map(JsonBundle::new)
+            .collect(Collectors.toList());
+    // this flag is used to ensure that we can iterate one more time after it was detected that a
+    // pass was successful such that we can do a final check that all is good
+    boolean again = false;
+
+    for (int i = ProfileMigratable.RETRY_COUNT; again || (i > 0); i--) {
```

Comment:
Based on the findings with the new logic of going with bundles first followed by features, I noticed a lot less retries required. I will therefore remove right away the ability to have multiple verification attempts in case the previous one did perform some operations. We should just be fine with the 5 attempts where the last one is just a verification one.

I also re-wrote the loop based on your face to face comments.